### PR TITLE
Add support to generate API token with `tanzu api-token create`

### PR DESCRIPTION
**Pull request by anujc25**
### What this PR does / why we need it
- Add support to generate API token with the `tanzu api-token create` command
- Add unit tests

Sample output:

```
> tz api-token create 
[i] Opening the browser window to complete the login 
Log in by visiting this link: https://wdc-10-206-98-55.eng.vmware.com/auth/oauth/authorize?client_id=tp_cli_app&code_challenge=XfCGpFSLGz5x_qYhXfsjYkXpfF6865RUh32wVd20-Ys&code_challe pe=code&state=1f5e2644c6e07bf7ff00cb5064bfabbe 

Optionally, paste your authorization code: [...] 

===

API Token Generation Successful! Your generated API token is: cle82767a13748debd53e33b2d46d92b-r

For non-interactive login use the API token as follows: TANZU_API_TOKEN=cle82767a13748debd53e33b2d46d92b-r tanzu login --endpoint https://wdc-10-206-98-55.eng.vmware.com

Please copy and save your token securely. Note that you will need to regenerate a new token before expiration time and login again to continue using the CLI.
```

### Which issue(s) this PR fixes

Fixes #

### Describe testing done for PR

- Manually tests the UAA login workflow. See the above screenshot

### Release note

```release-note
Add support to generate API token with `tanzu api-token create` (supported for UAA as IDP)
```

### CODE REVIEW

File: pkg/command/apitoken.go
Comment (vuil) [inline pkg/command/apitoken.go:64]: I think we need a message more comprehensible to user not familiar with how auth is implemented in TP. Maybe something to the effect that "The generation of the API TOKEN for <public endpoint> should be done via <link to CSP UI>")

File: pkg/command/apitoken.go
Comment (vuil) [inline pkg/command/apitoken.go:84]: This is not quite right, unfortunately ExpiresIn only refers to validity of the access token. I suggest we remove this line,
then L86 should be next, followed by L85, which needs to be edited to remove reference to the expiration time.

**Review by vuil — COMMENTED** [review]
Some comments about the output as I continue the review

File: pkg/utils/time/time.go
Comment (vuil) [inline pkg/utils/time/time.go:9]: had to do something similar in this [PR](https://github.com/vmware-tanzu/tanzu-cli/pull/819/files#diff-e7351021bcc81adaa7ed3b1f735852841a7bad195f45b957c71aca058f04d9a4R23), but putting this in its own package is cleaner.

File: pkg/command/apitoken.go
Comment (vuil) [inline pkg/command/apitoken.go:72]: reminder for one of the two PRs to leverage https://github.com/vmware-tanzu/tanzu-cli/pull/819/files#diff-21acb112c12bd989b0b1cc0d898ba6479f1f7c5c914fc9baf5b9500d65c7833eR28

File: pkg/command/apitoken_test.go
Comment (vuil) [inline pkg/command/apitoken_test.go:179]: Should we remember the uaa.TanzuLogin and restore it on a defer so we minimize the impact of the function patching?

**Review by vuil — COMMENTED** [review]
looking good in general, thanks the for changes.
I have a few questions and suggestions.
And some UX output needs to be updated

File: pkg/command/apitoken.go
Comment (marckhouzam) [inline pkg/command/apitoken.go:55]: Nit: Error message normally don't start with an uppercase letter.  We used to get a warning about these.  I'm not sure why we don't anymore.  Unless I remember the warning from another repo...

File: pkg/command/apitoken.go
Comment (marckhouzam) [inline pkg/command/apitoken.go:58]: Same nit about uppercase first letter

File: pkg/command/apitoken.go
Comment (marckhouzam) [inline pkg/command/apitoken.go:61]: Nit: can we use `config.UAAIdpType` instead of `"uaa"`?

File: pkg/command/apitoken.go
Comment (marckhouzam) [inline pkg/command/apitoken.go:82]: "follow" -> "follows"

File: pkg/command/apitoken.go
Comment (marckhouzam) [inline pkg/command/apitoken.go:82]: Just a thought; would it be nicer for the user if we printed the real token instead `<token>` in the above instructions?

Comment (marckhouzam) [inline pkg/command/apitoken.go:82]: This token can only be used for UAA, correct?
If so, instead of `<tanzu-platform-endpoint>`, should we put something with `uaa` in it?

File: pkg/command/apitoken.go
Comment (marckhouzam) [reply pkg/command/apitoken.go]: Do we know the expiration time easily?
Since we tell the user `regenerate a new token before expiration`, I find it nice to tell them when that will be.  If not, is there another way for them to avoid their usage failing when expiration happens?

**Review by marckhouzam — COMMENTED** [review]
Nice.
Minor suggestions to take or leave.

File: pkg/command/apitoken.go
Comment (vuil) [reply pkg/command/apitoken.go:82]: That UAA is used is implementation details that may be lost on many users. In a comment earlier, I suggested using terminology that is more comprehensible. Still thinking of options...

**Comment by anujc25** [comment]
> Do we know the expiration time easily?
Since we tell the user regenerate a new token before expiration, I find it nice to tell them when that will be. If not, is there another way for them to avoid their usage failing when expiration happens?

@marckhouzam regarding this, I discussed with @vuil and the conclusion is we do not have a way to know the expiration time for the Refresh Token at the moment. So, if needed we need to document the expiration time separately in the docs.

**Comment by marckhouzam** [comment]
> @marckhouzam regarding this, I discussed with @vuil and the conclusion is we do not have a way to know the expiration time for the Refresh Token at the moment. So, if needed we need to document the expiration time separately in the docs.

So we will document that the token has a lifetime of X time?
Do we prefer only putting it in the docs so we can more easily change it, instead of printing the value we currently know?

I assume there is no advanced notification for the user to know the token is about to expire?

**Comment by anujc25** [comment]
> > @marckhouzam regarding this, I discussed with @vuil and the conclusion is we do not have a way to know the expiration time for the Refresh Token at the moment. So, if needed we need to document the expiration time separately in the docs.
> 
> So we will document that the token has a lifetime of X time? Do we prefer only putting it in the docs so we can more easily change it, instead of printing the value we currently know?
> 
> I assume there is no advanced notification for the user to know the token is about to expire?

Yes. We do not know the lifetime of this refresh token at the moment. And that can be updated without CLI's knowledge from the backend so it is safe to put that information in the docs.
Correct, there isn't any advanced notification for the user to know the token is about to expire.
Please note that this API token generation from the CLI is a temporary workaround for supporting the non- interactive login with UAA because UAA does not have any UI to generate this token. There will be updates and enhancement in this area in future would likely make CLI more user friendly.

File: pkg/command/apitoken.go
Comment (marckhouzam) [reply pkg/command/apitoken.go]: Nice !

Comment (marckhouzam) [reply pkg/command/apitoken.go]: You are right.  Maybe "TPSM" is more appropriate, or the long form of it?

File: pkg/command/apitoken.go
Comment (anujc25) [reply pkg/command/apitoken.go]: Apologies for the confusion and not using this thread to discuss this topic. I responded separately https://github.com/vmware-tanzu/tanzu-cli/pull/820#issuecomment-2397636551

Also, I have updated the PR description to remove the outdated screenshot and added the sample output (without color).

File: pkg/auth/uaa/tanzu.go
Comment (marckhouzam) [inline pkg/auth/uaa/tanzu.go:21]: This is the only thing that prevents me from approving.  Unless we want to merge and update this when the info is available?

File: pkg/utils/time/time.go
Comment (marckhouzam) [inline pkg/utils/time/time.go:9]: I don't see this used anywhere.
Also, I don't understand how it would be used...  What is the different in calling `time.Now()` from go and `time.Now()` from this new package?

File: pkg/auth/common/login_handler.go
Comment (marckhouzam) [inline pkg/auth/common/login_handler.go:119]: I see in @vuil PR https://github.com/vmware-tanzu/tanzu-cli/pull/819/files#diff-c2506254f4d5d93b4c84bfa32c18782a79fbca5c788f38444419c9af5060cca1R130 that he also sets the `h.oauthConfig.ClientID` field here.

File: pkg/utils/time/time.go
Comment (anujc25) [reply pkg/utils/time/time.go:9]: I will remove this now. I used in the earlier version of this PR for unit tests but I removed that because we removed the expiration time from the log message.

File: pkg/auth/common/login_handler.go
Comment (anujc25) [reply pkg/auth/common/login_handler.go:119]: This needs to be updated as part of rebase after #819 gets merged because the same function is implemented there as well.

File: pkg/utils/time/time.go
Comment (anujc25) [reply pkg/utils/time/time.go:9]: I will keep this for now and update the https://github.com/vmware-tanzu/tanzu-cli/pull/819#discussion_r1790783947 as part of the rebase.

File: pkg/command/apitoken.go
Comment (anujc25) [reply pkg/command/apitoken.go]: Updated to use `uaa.GetAlternateClientID()` after rebase.

File: pkg/auth/uaa/tanzu.go
Comment (anujc25) [reply pkg/auth/uaa/tanzu.go]: Rebased. This does not apply anymore.

File: pkg/command/apitoken.go
Comment (vuil) [reply pkg/command/apitoken.go]: Maybe the way it is implemented is reasonable enough. Users who have to use the form "tanzu login --endpoint ..." and who has successfully use "api-token create" to get to this point should know what tanzu-platform-endpoint refers to.

Comment (anujc25) [reply pkg/command/apitoken.go]: I have updated the log message to mention actual Tanzu Platform endpoint that needs to be used instead of `<tanzu-platform-endpoint>`. Let me know if any other changes are needed here.

File: pkg/auth/common/token.go
Comment (vuil) [inline pkg/auth/common/token.go:102]: thanks :-)

File: pkg/command/apitoken.go
Comment (vuil) [inline pkg/command/apitoken.go:71]: Suggest: remove reference to UAA IDP type. 
maybe

Command no supported. Please refer to documentation on how to generate an API token for a public Tanzu Platform endpoint via https://console.tanzu.broadcom.com

File: pkg/command/apitoken.go
Comment (vuil) [inline pkg/command/apitoken.go:100]: We may have to workshop the err/output messages in a followup. Some thoughts


- Is it better to output the token in stdout, and show the rest in stderr?
- do we want to advocate TANZU_API_TOKEN=%s tanzu login, or mention the config file way to persist the env var?

The messages in these function

- may confuse users who are interacting with the public endpoint, who ends up creating a valid `tanzu` endpoint and still get a surprise when this fails for him
- the instructions for regenerate and re-login might be confusing to the user of the current CLI instance (who has obviously logged in)

File: docs/cli/commands/tanzu_api-token_create.md
Comment (anujc25) [inline docs/cli/commands/tanzu_api-token_create.md:18]: extra spaces at the front.

**Review by vuil — APPROVED** [review]
I think we may still want to tweak the messages but the rest lgtm, thanks!

File: pkg/command/apitoken.go
Comment (anujc25) [reply pkg/command/apitoken.go:100]: > Is it better to output the token in stdout, and show the rest in stderr?

I would prefer using `stdout` for both the token and the rest of the output. Since this is an output message intended for the user, it's more conventional to keep all output in `stdout`. 

> do we want to advocate TANZU_API_TOKEN=%s tanzu login, or mention the config file way to persist the env var?

I feel persisting the env var in the config file is a generic way to configure any env variable. So I don't feel to strongly that we also mention the persisting the TANZU_API_TOKEN in the configuration file. Also if user is not aware of what it means to persist a setting in the config file as env.TANZU_API_TOKEN it might create even more confusion because all the follow up `tanzu login` attempt will try to use the token automatically.

**Pull request merged**
